File: src/main/java/no/fintlabs/consumer/admin/AdminController.java
package no.fintlabs.consumer.admin;

import lombok.RequiredArgsConstructor;
import no.fintlabs.cache.CacheService;
import no.fintlabs.consumer.config.ConsumerConfiguration;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.stream.Collectors;

import static no.fintlabs.consumer.config.EndpointsConstants.ADMIN;

@RestController
@RequestMapping(ADMIN)
@RequiredArgsConstructor
public class AdminController {

    private final CacheService cacheService;
    private final ConsumerConfiguration configuration;

    @GetMapping("/health")
    public ResponseEntity<?> getHealthChecks() {
        // TODO: Implement when status service is working!
        return null;
    }

    @Deprecated
    @GetMapping({"/organisations"})
    public Collection<String> getOrganisations() {
        return new ArrayList<>();
    }

    @Deprecated
    @GetMapping("/organisations/{orgId:.+}")
    public Collection<String> getOrganization(@PathVariable String orgId) {
        return new ArrayList<>();
    }

    @GetMapping("/assets")
    public Collection<String> getAssets() {
        return new HashSet<>(List.of(configuration.getOrgId()));
    }

    @Deprecated
    @GetMapping("/caches")
    public Map<String, Integer> getCaches() {
        return new HashMap<>();
    }

    @GetMapping("/cache/status")
    public Map<String, CacheEntry> getCacheStatus() {
        return cacheService.getResourceCaches().entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        entry -> new CacheEntry(new Date(entry.getValue().getLastUpdated()), entry.getValue().size())
                ));
    }

    @PostMapping({"/cache/rebuild", "/cache/rebuild/{model}"})
    public void rebuildCache(
            @RequestHeader(name = "x-client") String client,
            @PathVariable(required = false) String model
    ) {
        // TODO: Yet to be implemented
    }

}

File: src/main/java/no/fintlabs/consumer/admin/AdminController.kt
package no.fintlabs.consumer.admin

import no.fintlabs.cache.CacheService
import no.fintlabs.consumer.config.ConsumerConfiguration
import no.fintlabs.consumer.config.EndpointsConstants
import no.fintlabs.consumer.kafka.sync.SyncTimestampStore
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.Date

@RestController
@RequestMapping(EndpointsConstants.ADMIN)
class AdminController(
    private val cacheService: CacheService,
    private val configuration: ConsumerConfiguration,
    private val syncTimestampStore: SyncTimestampStore,
) {
    @GetMapping("/health")
    fun healthChecks(): ResponseEntity<*>? = null // TODO: Implement when status service is working!

    @GetMapping("/organisations")
    @Deprecated("")
    fun organisations(): MutableCollection<String> = ArrayList<String>()

    @Deprecated("")
    @GetMapping("/organisations/{orgId:.+}")
    fun getOrganization(
        @PathVariable orgId: String?,
    ): MutableCollection<String?> {
        return ArrayList<String?>()
    }

    @GetMapping("/assets")
    fun assets(): MutableCollection<String> = hashSetOf(configuration.orgId.value)

    @GetMapping("/caches")
    @Deprecated("")
    fun caches(): MutableMap<String, Int> = HashMap<String, Int>()

    /**
     * Get status for all caches.
     *
     * @return an object where each key is a resource name and each value is an object containing
     * lastUpdated and size for the cache for that resource name
     */
    @GetMapping("/cache/status")
    fun cacheStatus(): Map<String, CacheEntry> =
        cacheService.getCachedResourceNames().associateWith { resourceName ->
            val cache = cacheService.getCache(resourceName)
            val lastFullSync = syncTimestampStore.getLastFullSync(resourceName)?.let { Date.from(it) }
            CacheEntry(Date(cache.lastUpdated), lastFullSync, cache.size)
        }

    @PostMapping("/cache/rebuild", "/cache/rebuild/{model}")
    fun rebuildCache(
        @RequestHeader(name = "x-client") client: String?,
        @PathVariable(required = false) model: String?,
    ) {
        // TODO: Yet to be implemented
    }
}
